### PR TITLE
add another serializable strategy

### PR DESCRIPTION
**Pull request by feeblefakie**
This PR contains updates for another serializable strategy proposed by @Yonezawa-T2  in
https://github.com/scalar-labs/scalardb/pull/70 .

There are still some unknowns as discussed in #70  but I think it is a good alternative so I actually implemented it as yet another `SerializableStrategy` and named it `Extra Read` strategy (the existing strategy is now called `Extra Write` strategy). We will do benchmarking and verification once it is merged.

### CODE REVIEW

File: src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java [hunk 1]
```diff
@@ -105,7 +116,9 @@ public void put(Snapshot.Key key, Delete delete) {
   }
 
   public void to(MutationComposer composer) {
-    if ((composer instanceof PrepareMutationComposer) && isolation == Isolation.SERIALIZABLE) {
+    if ((composer instanceof PrepareMutationComposer)
+        && isolation == Isolation.SERIALIZABLE
+        && strategy == SerializableStrategy.EXTRA_WRITE) {
       toSerializable();
```
Comment (feeblefakie) [inline src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:122]: This method will be refactored in another PR by having `toSerializableWithExtraWrite` and the above if statement will be in the method for consistency with `toSerializableWithExtraRead`.

File: src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitManager.java [hunk 2]
```diff
@@ -62,15 +62,39 @@ public ConsensusCommit start(String txId) {
 
   @Override
   public synchronized ConsensusCommit start(Isolation isolation) {
-    String txId = UUID.randomUUID().toString();
-    return start(txId, isolation);
+    return start(isolation, SerializableStrategy.EXTRA_WRITE);
```
Comment (feeblefakie) [inline src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitManager.java:65]: After some benchmarking and verification, we can change the default strategy based on the results.

File: src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java [hunk 3]
```diff
@@ -164,6 +178,71 @@ void toSerializable() {
...
+        if (transactionResult.getId().equals(id)) {
+          continue;
+        }
+        currentReadSetByScan.add(transactionResult);
+      }
+
+      for (Key key : entry.getValue().get()) {
+        if (writeSet.containsKey(key)) {
+          continue;
+        }
```
Comment (Yonezawa-T2) [inline src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:204]: Should we also check delete-set?

Comment (feeblefakie) [reply src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:204]: Good catch. Yes, it should be checked as well. Thank you.

Comment (feeblefakie) [reply src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java]: Fixed in cdb12d1.
